### PR TITLE
fixed unescaping of graphvis escString (bsc#1157916)

### PR DESCRIPTION
**Pull request by aschnell**
In the graphviz file a escString can be used for label and tooltip. The escString can include special substrings, e.g. ```\n``` or ```\\```. Those where not handled properly which resulted in the verbatim escString ```\\x``` being displayed as ```\\x``` instead of ```\x``` as required.

Part of https://trello.com/c/uv6c4Y1p/1333-2-p3-1135341-yast2-disk-coredump-problem-parsing-subvolume-name.

Also see https://bugzilla.suse.com/show_bug.cgi?id=1157916.

### CODE REVIEW

File: src/QY2Graph.cc [hunk 1]
```diff
@@ -442,6 +439,51 @@ QY2Graph::renderGraph(graph_t* graph)
...
+		case 'n':
+		case 'l':
+		case 'r':
+		    r += '\n';
+		    break;
+
+		case '\\':
+		    r += '\\';
+		    break;
+	    }
```
Comment (mvidner) [inline src/QY2Graph.cc]: What about the `default` case?

https://graphviz.gitlab.io/_pages/doc/info/attrs.html#k:escString says:
>  In addition, if the associated attribute is label, headlabel or taillabel, the escape sequences "\n", "\l" and "\r" divide the label into lines, centered, left-justified, and right-justified, respectively.
> 
> Obviously, one can use "\\" to get a single backslash. *A backslash appearing before any character not listed above is ignored.*

So we should implement it with
```c++
		default:
		    r += c;
		    break;
```
That will also render the "label macros" like \G \N \E literally but that's fine, IMHO better than dropping them like now.

File: src/QY2Graph.cc [hunk 2]
```diff
@@ -442,6 +439,51 @@ QY2Graph::renderGraph(graph_t* graph)
...
+		    r += c;
+		    break;
+	    }
+	}
+	else
+	{
+	    switch (c)
+	    {
+		case 'n':
+		case 'l':
```
Comment (mvidner) [inline src/QY2Graph.cc]:
```suggestion
		case 'l': // newline, align Left
```

File: src/QY2Graph.cc [hunk 3]
```diff
@@ -442,6 +439,51 @@ QY2Graph::renderGraph(graph_t* graph)
...
+		case 'r':
```
Comment (mvidner) [inline src/QY2Graph.cc]:
```suggestion
		case 'r': // newline, align Right
```

File: package/libyui-qt-graph.changes [hunk 4]
```diff
@@ -1,3 +1,9 @@
+-------------------------------------------------------------------
+Thu Nov 28 12:51:31 CET 2019 - aschnell@suse.com
+
+- Fixed unescaping of graphviz escString (bsc#1157916)
```
Comment (mvidner) [inline package/libyui-qt-graph.changes]: That is correct but requires the reader to perform a web lookup of "graphviz escString".
Let's spill the details right away:

```suggestion
- Backslash escapes in labels are respected now
  (graphviz escString, bsc#1157916)
```

File: src/QY2Graph.cc (hunk 1 above)
Comment (aschnell) [reply src/QY2Graph.cc]: AFAIS a default case would hurt here, the doc says "A backslash appearing before any character not listed above is ignored."

Comment (mvidner) [reply src/QY2Graph.cc]: Ah, I see. Your interpretation is "the backslash, and the following character, are ignored".
My interpretation was "only the backslash is ignored, the following character is rendered'.
It turns out graphviz follows my interpretation. Try
```sh
echo 'digraph g { "foo\t"; }' | dotty -
```

File: src/QY2Graph.cc [hunk 5]
```diff
@@ -442,6 +439,52 @@ QY2Graph::renderGraph(graph_t* graph)
 }
 
 
+QString
+QY2Graph::unescape(const std::string& s) const
+{
+    std::string r;
```
Comment (lslezak) [inline src/QY2Graph.cc]: Just some questions: How often an escape is really used? How long the strings usually are?

We are just creating a duplicate string one-by-one character. Maybe we could do something like

```c++
if (s.find('\\') == std::string::npos)
  return QString::fromUtf8(s.c_str())
```

to optimize it for not escaped strings. But if the strings are small this optimization is probably not needed...

Comment (aschnell) [reply src/QY2Graph.cc]: In the use case I know of the strings are usually not so long. Tooltips strings might be a bit longer but usually also include newlines.

In general I think any optimization for speed here is not needed.

**Pull request merged**